<a href="https://colab.research.google.com/github/sam69387/gensim/blob/master/Yang_Shi_Lab1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DMA Assignment 1 - Data Preprocessing

Data transformations are useful for preparing a dataset for answering a particular question. Part of this process involves generating features from the dataset you find relevant to the question at hand. For this lab, we will be using a Yelp reviews dataset. Each row in the dataset depicts one review along with the features of the review (the reviewer, the review text, etc.). The goal of this lab is to convert this reviews dataset into a reviewers dataset by creating different features of each reviewer.

The submission for this assignment should be done *individually*, although you are allowed to work in groups of 2.



# Environment Setup
Run this cell to setup your environment.

In [0]:
# Importing libraries
import pandas as pd
import math
import numpy as np
print('Libraries Imported')

#DOWNLOADING DATASET
!wget http://people.ischool.berkeley.edu/~zp/course_datasets/yelp_reviews.csv
#!unzip yelp_reviews.zip
print('Dataset Downloaded: yelp_reviews.csv')
df=pd.read_csv('yelp_reviews.csv')
print(df.head())

print('Setup Complete')

Libraries Imported
--2019-02-07 17:05:08--  http://people.ischool.berkeley.edu/~zp/course_datasets/yelp_reviews.csv
Resolving people.ischool.berkeley.edu (people.ischool.berkeley.edu)... 128.32.78.16
Connecting to people.ischool.berkeley.edu (people.ischool.berkeley.edu)|128.32.78.16|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 376638166 (359M) [text/csv]
Saving to: ‘yelp_reviews.csv.6’

yelp_reviews.csv.6  100%[===================>] 359.19M  1.81MB/s    in 84s     

2019-02-07 17:06:31 (4.29 MB/s) - ‘yelp_reviews.csv.6’ saved [376638166/376638166]

Dataset Downloaded: yelp_reviews.csv
     type             business_id                 user_id  stars  \
0  review  mxrXVZWc6PWk81gvOVNOUw  mv7shusL4Xb6TylVYBv4CA      4   
1  review  mxrXVZWc6PWk81gvOVNOUw  0aN5QPhs-VwK2vusKG0waQ      5   
2  review  kK4AzZ0YWI-U2G-paAL7Fg  0aN5QPhs-VwK2vusKG0waQ      5   
3  review  mxrXVZWc6PWk81gvOVNOUw  1JUwyYab-uJzEx_FRd81Zg      5   
4  review  mxrXVZWc6PWk81gvOVNOUw  2Zd3

**Q1: What was the highest number of reviews for any one business_id?**
- For this task, we will need to group the reviews dataset by business_id. This will aggregate data for each business, which is what we need. This can be done using [groupby](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.groupby.html).
    - ```yelp_businesses = yelp_dataset.groupby('business_id').size() ```
    - The .size() counts the number of instances for each business_id, which gives us the number of reviews as each instance in this dataset is a review.
    - The following command will sort this list, after which you can take note of the highest value: ```sorted_yelp_businesses = yelp_businesses.sort(ascending=False,inplace=False)```
    - This approach allows you to see the data structure being used in the sort. A quicker approach to getting the max would be to use the max function: ```max(yelp_businesses)```
<br>
<br>

In [0]:
df.columns

Index([u'type', u'business_id', u'user_id', u'stars', u'text', u'date',
       u'cool_votes', u'useful_votes', u'funny_votes'],
      dtype='object')

In [0]:
grouped_reviews = df.groupby('business_id').size()
max(grouped_reviews)

4128

###Answer:4128

**Q2: What was the average number of reviews for a business_id?**

In [0]:
grouped_reviews.mean()

12.63413902163123

###Answer:12.63413902163123

**Q3: What is the average number of reviews per reviewer?**

In [0]:
grouped_reviewers = df.groupby('user_id').size()
grouped_reviewers.mean()

3.188511934933203

###Answer:3.188511934933203

**Q4: What is the average number of cool votes per reviewer?**

In [0]:
grouped_reviewers_sum = df.groupby('user_id').sum()
grouped_reviewers_sum.mean()['cool_votes']

1.2417282785380945

###Answer:1.2417282785380945

**Q5: What is the average number of funny votes per reviewer?**

In [0]:
grouped_reviewers_sum.mean()['funny_votes']

1.10126486404605

###Answer:1.10126486404605



**Q6: What is the average number of useful votes per reviewer?**

In [0]:
grouped_reviewers_sum.mean()['useful_votes']

2.484476138872867

###Answer:2.484476138872867

**Q7: What is the average of the log of the number of reviews per reviewer?**

In [0]:
grouped_reviewers.head()
np.log(grouped_reviewers).mean()

0.7681766036714884

###Answer:0.7681766036714884

**Q8: What is the average of the log of the number of cool votes per reviewer?**

In [0]:
np.log(grouped_reviewers_sum)['cool_votes'].mean()

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:1: RuntimeWarning: divide by zero encountered in log
  """Entry point for launching an IPython kernel.


-inf

###Answer:-inf

**Q9: What is the average of the log of the number of funny votes per reviewer?**

In [0]:
np.log(grouped_reviewers_sum)['funny_votes'].mean()

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:1: RuntimeWarning: divide by zero encountered in log
  """Entry point for launching an IPython kernel.


-inf

###Answer:-inf

**Q10: What is the average of the log of the number of useful votes per reviewer?**

In [0]:
np.log(grouped_reviewers_sum)['funny_votes'].mean()


/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:1: RuntimeWarning: divide by zero encountered in log
  """Entry point for launching an IPython kernel.


-inf

###Answer:-inf

**Q11: Find the average of the percentage of total cool votes out of total votes for each reviewer.**

In [0]:
grouped_reviewers_sum['total_votes'] = np.sum(grouped_reviewers_sum.loc[:,'cool_votes':'funny_votes'],axis=1)
grouped_reviewers_sum['cool_percent'] = grouped_reviewers_sum['cool_votes']/grouped_reviewers_sum['total_votes']
'{:.4%}'.format(grouped_reviewers_sum.mean()['cool_percent'])

'19.2704%'

###Answer:19.2704%

**Q12: Find the average of the percentage of total funny votes out of total votes for each reviewer.**

In [0]:
grouped_reviewers_sum['funny_percent'] = grouped_reviewers_sum['funny_votes']/grouped_reviewers_sum['total_votes']
'{:.4%}'.format(grouped_reviewers_sum.mean()['funny_percent'])


'18.2564%'

###Answer:18.2564%

**Q13: Find the average of the percentage of total useful votes out of total votes for each reviewer.**

In [0]:
grouped_reviewers_sum['useful_percent'] = grouped_reviewers_sum['useful_votes']/grouped_reviewers_sum['total_votes']
'{:.4%}'.format(grouped_reviewers_sum.mean()['useful_percent'])


'62.4732%'

Answer:62.4732%

**Q14: Average review text length (in non-space characters)**

In [0]:
df['text_length'] = df['text'].str.len() - df['text'].str.count(' ')
df['text_length'].mean()

499.18610090393645

Answer: 499.18610090393645

**Q15: Year in which the reviewer wrote the most reviews. Once you have this for each reviewer, subtract the minimum possible year (2005) from each so that your final feature values are 0, 1, 2 etc.**

In [0]:
df['date'] = pd.to_datetime(df['date'], format = '%Y-%m-%d')
df['year'] = df['date'].dt.year
df['year_gap'] = df['year'] - 2005

In [0]:
df['length_ranked'] = df.groupby('user_id')['text_length'].rank(ascending=False)
df.sort_values('length_ranked', ascending=False).drop_duplicates('user_id').sort_index().head()

type             business_id                 user_id  stars  \
0   review  mxrXVZWc6PWk81gvOVNOUw  mv7shusL4Xb6TylVYBv4CA      4   
2   review  kK4AzZ0YWI-U2G-paAL7Fg  0aN5QPhs-VwK2vusKG0waQ      5   
3   review  mxrXVZWc6PWk81gvOVNOUw  1JUwyYab-uJzEx_FRd81Zg      5   
18  review  Cr2JNnewX53TYd9w6qw6Jw  2Zd3Xy8hUVmZkNg7RyNjhg      4   
24  review  IwShjaQLAvqV2msU4bh3ww  fuGfZWfDFf97d6UyuO3d8w      5   

                                                 text       date  cool_votes  \
0   Definitely try the duck dish.   I rank it amon... 2011-06-13           0   
2              Unbelievable sandwiches! Good service. 2011-06-25           0   
3   Awesome, awesome, awesome! My mom and sister a... 2011-07-18           1   
18                            Good burger good value. 2011-07-26           0   
24  This place is a Gem! It's small, dont get me w... 2013-11-28           0   

    useful_votes  funny_votes  text_length  year  year_gap  length_ranked  
0              0            0          162  2011         6            1.0  
2              0            0           35  2011         6            2.0  
3              1            0          257  2011         6            1.0  
18             1            0           20  2011         6           17.0  
24             0            0          108  2013         8            4.0

###Answer:[6, 9, 3, 11, 11, ..., 9, 10, 10, 11]

**Q16: Come up with a new feature. This may be derived from existing features. Give your feature the name ** *my_new_feature.* **Dsiplay head() of this new feature.**

In [0]:
df['5-star review'] = np.where(df['stars']==5,'Yes','No')
df['5-star review'].head()

0     No
1    Yes
2    Yes
3    Yes
4     No
Name: 5-star review, dtype: object

Answer: Yes / No